symengine/polys.cpp

```
@@ -0,0 +1,348 @@
+{
+    for(auto &a: set) {
+        int count = 0;
+        for(auto &b: set) {
+            if (b.first == a.first)
+                count++;
+            if (count == 2)
+                return false;
+        }
+    }
+    return true;
+}
+
+std::size_t Polynomial::__hash__() const
+{
+    std::hash<std::string> hash_string;
+    std::size_t seed = POLYNOMIAL;
+
+    for (auto &it: vars_)
+        seed += hash_string(it->get_name());
+
+    for (auto &it : polys_set_)
+    {
+        std::size_t temp = POLYNOMIAL;
+        hash_combine<long long>(temp, it.first);
+#ifdef HAVE_SYMENGINE_PIRANHA
+        hash_combine<long long>(temp, mpz_get_si(it.second.get_mpz_view()));
+#else
+        hash_combine<long long>(temp, (it.second).get_ui());
+#endif
+        seed += temp;
+    }
+    return seed;
+}
+
+bool Polynomial::__eq__(const Basic &o) const
+{
+    const Polynomial &s = static_cast<const Polynomial &>(o);
+
+    for (unsigned int i = 0; i < vars_.size(); i++) {
+        if (!vars_[i]->__eq__(*s.vars_[i]))
+            return false;
+    }
+
+    #ifdef HAVE_SYMENGINE_PIRANHA
+    if (hash_set_eq(polys_set_, s.polys_set_))
+        return true;
+    #else
+    if (unordered_set_eq(polys_set_, s.polys_set_))
+        return true;
+    #endif
+
+    return false;
+}
+
+int Polynomial::compare(const Basic &o) const
+{
+    const Polynomial &s = static_cast<const Polynomial &>(o);
+
+    if (polys_set_.size() != s.polys_set_.size())
+        return (polys_set_.size() < s.polys_set_.size()) ? -1 : 1;
+
+    for (unsigned int i = 0; i < vars_.size(); i++) {
+        int cmp = vars_[i]->compare(*s.vars_[i]);
+        if (cmp != 0)
+            return cmp;
+    }
+
+    #ifdef HAVE_SYMENGINE_PIRANHA
+    return hash_set_compare(polys_set_, s.polys_set_);
+    #else
+    throw::std::runtime_error("Not Implemented");
+    #endif
+}
+
+//TODO
+vec_basic Polynomial::get_args() const
+{
+    vec_basic args;
+    return args;
```

This should raise an exception before it gets implemented.